### ISSUE TITLE
Generate API from swagger spec

### ISSUE THREAD

**Issue by dnephin** [docker/docker#27919]
Related: https://github.com/docker/engine-api/issues/77, #5893, #22931

We now have a [swagger spec](https://github.com/docker/docker/blob/master/api/swagger.yaml) which we use to generate a few types. This issue is to document and track the rest of the required work.

## Project

* [ ] add a validate script to check that running the generation does not result in file changes (same idea as `validate-vendor`)
* [ ] replace [api reference docs](https://docker.github.io/engine/reference/api/docker_remote_api/) with [generated docs](https://bfirsh.github.io/docker-api-reference/). See https://github.com/bfirsh/docker-api-reference for a working prototype.
* [ ] document how contributors can work with the swagger spec and add new types.

## Milestone 1 - Types
 
Generate `api/types` from the spec:

**api/types/types.go**
* [ ] ContainerChange
* [ ] ImageHistory
* [ ] ImageDelete
* [ ] GraphDriverData
* [ ] RootFS
* [ ] ImageInspect
* [ ] Container
* [ ] CopyConfig
* [ ] ContainerPathStat
* [ ] ContainerStats
* [ ] ContainerProcessList
* [ ] Version
* [ ] Info
* [ ] PluginsInfo
* [ ] ExecStartCheck
* [ ] HealthcheckResult
* [ ] Health
* [ ] ContainerState
* [ ] ContainerNode
* [ ] ContainerJSONBase
* [ ] ContainerJSON
* [ ] NetworkSettings
* [ ] SummaryNetworkSettings
* [ ] NetworkSettingsBase
* [ ] DefaultNetworkSettings
* [ ] MountPoint
* [ ] NetworkResource
* [ ] EndpointResource
* [ ] NetworkCreate
* [ ] NetworkCreateRequest
* [ ] NetworkCreateResponse
* [ ] NetworkConnect
* [ ] NetworkDisconnect
* [ ] Checkpoint
* [ ] Runtime
* [ ] DiskUsage
* [ ] ImagesPruneConfig
* [ ] ContainersPruneConfig
* [ ] VolumesPruneConfig
* [ ] NetworksPruneConfig
* [ ] ContainersPruneReport
* [ ] VolumesPruneReport
* [ ] ImagesPruneReport
* [ ] NetworksPruneReport

**api/types/swarm/**:
* [ ] Version
* [ ] Meta
* [ ] Annotations
* [ ] ContainerSpec
* [ ] Endpoint
* [ ] EndpointSpec
* [ ] ResolutionMode
* [ ] PortConfig
* [ ] PortConfigProtocol
* [ ] EndpointVirtualIP
* [ ] Network
* [ ] NetworkSpec
* [ ] NetworkAttachmentConfig
* [ ] NetworkAttachment
* [ ] IPAMOptions
* [ ] IPAMConfig
* [ ] Driver
* [ ] Node
* [ ] NodeSpec
* [ ] NodeRole
* [ ] NodeAvailability
* [ ] NodeDescription
* [ ] Platform
* [ ] EngineDescription
* [ ] PluginDescription
* [ ] NodeStatus
* [ ] Reachability
* [ ] ManagerStatus
* [ ] NodeState
* [ ] Service
* [ ] ServiceSpec
* [ ] ServiceMode
* [ ] UpdateState
* [ ] UpdateStatus
* [ ] ReplicatedService
* [ ] GlobalService
* [ ] UpdateConfig
* [ ] ClusterInfo
* [ ] Swarm
* [ ] JoinTokens
* [ ] Spec
* [ ] OrchestrationConfig
* [ ] TaskDefaults
* [ ] RaftConfig
* [ ] DispatcherConfig
* [ ] CAConfig
* [ ] ExternalCAProtocol
* [ ] ExternalCAProtocol
* [ ] ExternalCA
* [ ] InitRequest
* [ ] JoinRequest
* [ ] LocalNodeState
* [ ] Info
* [ ] Peer
* [ ] UpdateFlags
* [ ] TaskState
* [ ] Task
* [ ] TaskSpec
* [ ] Resources
* [ ] ResourceRequirements
* [ ] Placement
* [ ] RestartPolicy
* [ ] RestartPolicyCondition
* [ ] TaskStatus
* [ ] ContainerStatus

## Milestone 2 - Client

Generate (most of) `client/` from the spec. 

This milestone can be done in parallel with Milestone 3, but requires a prototype before work can be enumerated.

## Milestone 3 - Server (router)

Generate (most of) `api/server/router/` from the spec. 

This milestone can be done in parallel with Milestone 2, but requires a prototype before work can be enumerated.

**Comment by stevvooe** [comment]
Have we considered generating the swagger definition from the API types? I don't see value in moving from a type-safe definition to a yaml-based language. In addition, now contributors that are familiar with Go must become familiar with swagger.

I thought the original goal was to have swagger documentation; not use swagger as a framework.

**Comment by dnephin** [comment]
> Have we considered generating the swagger definition from the API types?

We have, and @bfirsh experimented with it some. I don't think it's the right approach.

> I don't see value in moving from a type-safe definition to a yaml-based language. 

This is no different from specifying things a grpc `.proto` file. The specification is rich enough that you do not lose any type safety when translating to go types.

> I thought the original goal was to have swagger documentation; not use swagger as a framework.

An interface definition language isn't nearly as valuable if you only use it for documentation. You still have all the downsides of maintaining documentation separate from the code. We aren't using any frameworks here, we're using swagger as an interface definition language.

**Comment by AndrewGuenther** [comment]
Would love to help get this all landed. What's the current status on this? Is there a good place I can help out?

**Comment by bfirsh** [comment]
Auto-generated API docs have now shipped.

https://docs.docker.com/engine/api/v1.25/

(Ping @dnephin - I no longer have permission to tick that box ;)

**Comment by dnephin** [comment]
@AndrewGuenther There is still lots of work to do. I've been slowly converting a few types at a time. I have a PR open that does `Version`, and a wip branch for `Info`. Otherwise there are still many types to convert.

There are a bunch of linked PRs that can give you a sense of what needs to be done to convert a type. I do something like this:

1. Add it to `hack/generate-swagger-api.sh` and run it to generate the struct
2. Compare the struct against the old struct and adjust the `swagger.yaml` as necessary until they match (except for name). Sometimes I will use the new name if it's more descriptive.
3. Remove the old struct
4. Update all the code that references that type to use the new name or import location

**Comment by krasi-georgiev** [comment]
@dnephin 
I work on #6889  
do you think it would be crazy to think that swagger could also generate the cli cobra commands ?
I would be interested to look into it if you think it would be a good idea.

**Comment by stevvooe** [comment]
> do you think it would be crazy to think that swagger could also generate the cli cobra commands ?

Please no.

**Comment by krasi-georgiev** [comment]
@stevvooe can't say I didn't expected that reaction :) 
that said, hate to see the same command descriptions maintained twice.

**Comment by stevvooe** [comment]
> hate to see the same command descriptions maintained twice.

This is by far the least evil present in the docker documentation. 

In fact, they really have different use cases. The cli stuff is what you see when using and there may be manpages with supplemental information. Putting them into a single basket and tieing to only what is in the cli definitions will hamstring the quality of manpage documentation.

I'll cite the different between the `docker run` manpage and the autogenerated version. The manually maintained manpage is much more thorough.

In practice, I am not sure if the pursuit of DRY doesn't make for more work and a poorer result. The best work seems to be done by those willing to get on the ground and they waste less time automating.

Look at this way: would you rather live in a manufactured home or a bespoke castle?

**Comment by danbeaulieu** [comment]
Hi, not sure if this is the intent of the yaml file but I just tried to generate a client in go and java based on the yaml file in master and both failed.

```
$ java -jar modules/swagger-codegen-cli/target/swagger-codegen-cli.jar generate    -i https://raw.githubusercontent.com/docker/docker/master/api/swagger.yaml -l go -o `pwd`/docker
```

results in:
```
Exception: null
	at io.swagger.codegen.DefaultGenerator.processOperation(DefaultGenerator.java:827)
	at io.swagger.codegen.DefaultGenerator.processPaths(DefaultGenerator.java:730)
	at io.swagger.codegen.DefaultGenerator.generateApis(DefaultGenerator.java:356)
	at io.swagger.codegen.DefaultGenerator.generate(DefaultGenerator.java:666)
	at io.swagger.codegen.cmd.Generate.run(Generate.java:234)
	at io.swagger.codegen.SwaggerCodegen.main(SwaggerCodegen.java:41)
Caused by: java.lang.NullPointerException
	at io.swagger.codegen.DefaultCodegen.isDataTypeBinary(DefaultCodegen.java:2590)
	at io.swagger.codegen.DefaultCodegen.fromParameter(DefaultCodegen.java:2541)
	at io.swagger.codegen.DefaultCodegen.fromOperation(DefaultCodegen.java:2121)
	at io.swagger.codegen.DefaultGenerator.processOperation(DefaultGenerator.java:774)
	... 5 more
```

There is more information that points to /containers/create: https://github.com/docker/docker/blob/master/api/swagger.yaml#L2774

This looks like it is related to https://github.com/swagger-api/swagger-codegen/issues/3225

It doesn't look like this is supported yet by either swagger or docker but I'm curious just in case I'm doing something wrong...

**Comment by boonkerz** [comment]
Same Issue here any news?

**Comment by masaeedu** [comment]
@thaJeztah @danbeaulieu There's some problems with the YAML spec. First is the fact that inline `allOf` is broken, as you linked to in https://github.com/swagger-api/swagger-codegen/issues/3225. This issue can be fairly easily resolved by moving all the inline `allOf`s out to the `definitions` key. I've opened #35132 to fix this.

The second issue is that in a couple places (e.g. `#/definitions/ContainerConfig/Cmd`), `type` is being defined as an array of types `["array", "string"]`, whereas `swagger` seems to only be happy with a single string for `type`. The schema validator chokes on these locations (you can test it out by pasting into https://editor.swagger.io).

It is a bit strange, because [JSON schema](https://cswr.github.io/JsonSchema/spec/multiple_types/) is perfectly happy with an array of types for `type`. Either the discrepancy between JSON schema and swagger-codegen/editor needs to be filed as a bug on their end, or this needs to be re-expressed in some form Swagger is happy with (presumably using `anyOf`/`oneOf`). I haven't been able to figure out the latter yet.

**Comment by masaeedu** [comment]
Turns out things are a little worse than that: I was looking at the wrong spec above. You're currently using [OpenAPI 2.0](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md), which doesn't support `anyOf`/`oneOf` either (in addition to not supporting `type: ["number", "string"]`). 

This means there's basically no way to describe a parameter that can be one of two different things in OpenAPI 2.0 (e.g. a string or an array of strings in the case of `Cmd`).

Until swagger-codegen gets support for OpenAPI 3.0, one way to compromise might be to only represent the more general of the two possibilities in the API spec. There's only two instances of this problem in swagger.yaml, and they both involve parameters that can be either a single string or an array of strings. You could just drop the single string overload in the API spec, and single strings could be ferried between the client and server as a single element string array.

**Comment by thaJeztah** [comment]
Yes, the `Cmd` is a bit odd; it looks like the client defaults to sending an array (?) or at least it's converted to an array by the daemon. Sending a string _does_ work though (likely for backward-compatibility reasons);

```bash
$ curl -v \
  --unix-socket /var/run/docker.sock \
  -X POST \
  "http://localhost/v1.32/containers/create?name=foo" \
  -H "Content-Type: application/json" \
  -d '{"Cmd":"sh -c \"echo foobar\"","Image":"busybox"}'

$ docker inspect --format '{{json .Config.Cmd}}' foo
["sh -c \"echo foobar\""]
```


```bash
$ docker run -dit --name foo2 busybox sh -c "echo hello foobar"
$ docker inspect --format '{{json .Config.Cmd}}' foo2
["sh","-c","echo hello foobar"]
```

**Comment by masaeedu** [comment]
@thaJeztah So what do you think, is it okay to just substitute `type: ["array", "string"]` with `type: "array"` here?

**Comment by thaJeztah** [comment]
Possibly yes; I'm leaning towards describing it as an array of strings, and in the description we can mention something like "For backward compatibility, this property also accepts a string" (or similar)

If you want to update that, please do so in a separate pull request, as it may need some discussion/input from other maintainers, and I'd like to prevent that discussion from blocking the other changes 👍

**Comment by masaeedu** [comment]
@thaJeztah See #35146.

**Comment by denibertovic** [comment]
What's the status of this issue?

**Comment by boonkerz** [comment]
We should move to version open api version 3 or?
